// @clickhouse/client-common/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClickHouseClient = void 0;
const parse_1 = require("./parse");
const logger_1 = require("./logger");
const config_1 = require("./config");
const tracing_1 = require("./tracing");
class ClickHouseClient {
    clientClickHouseSettings;
    connectionParams;
    connection;
    makeResultSet;
    valuesEncoder;
    sessionId;
    role;
    jsonHandling;
    tracer;
    constructor(config) {
        const logger = config?.log?.LoggerClass
            ? new config.log.LoggerClass()
            : new logger_1.DefaultLogger();
        const configWithURL = (0, config_1.prepareConfigWithURL)(config, logger, config.impl.handle_specific_url_params ?? null);
        this.connectionParams = (0, config_1.getConnectionParams)(configWithURL, logger);
        this.clientClickHouseSettings = this.connectionParams.clickhouse_settings;
        this.sessionId = config.session_id;
        this.role = config.role;
        this.connection = config.impl.make_connection(configWithURL, this.connectionParams);
        // Using the connection params log level as it does the parsing.
        // TODO: it would be better to parse the log level in the client itself.
        this.makeResultSet = config.impl.make_result_set;
        this.jsonHandling = {
            ...parse_1.defaultJSONHandling,
            ...config.json,
        };
        this.valuesEncoder = config.impl.values_encoder(this.jsonHandling);
        // Assigned once at client creation: when no tracer is configured, the
        // shared no-op tracer keeps the hot path branch-free.
        this.tracer = config.tracer ?? tracing_1.NoopClickHouseTracer;
    }
    /**
     * Used for most statements that can have a response, such as `SELECT`.
     * Returns an implementation of {@link BaseResultSet}.
     *
     * The `FORMAT` clause should be specified separately via {@link QueryParams.format} (default is `JSON`);
     * this method will always append `FORMAT <format>` to the end of {@link QueryParams.query}.
     * If the query already contains a `FORMAT` clause, ClickHouse will return a syntax error due to a duplicate `FORMAT`.
     * This is intended behavior.
     * Use {@link ClickHouseClient.insert} for data insertion, {@link ClickHouseClient.command} for DDLs,
     * or {@link ClickHouseClient.exec} for queries where you need to provide the full SQL (including `FORMAT`) yourself or where the `FORMAT` suffix is not supported.
     *
     * @note For `SHOW [ROW] POLICIES`, use the full syntax `SHOW POLICIES ON *`,
     * as the short version does not support appending `FORMAT` at the server SQL parser level.
     * See https://github.com/ClickHouse/ClickHouse/issues/105899
     *
     * See {@link DataFormat} for the formats supported by the client.
     */
    async query(params) {
        const format = params.format ?? "JSON";
        const query = formatQuery(params.query, format);
        const queryParams = this.withClientQueryParams(params);
        const { log_writer, log_level } = this.connectionParams;
        return this.tracer.startActiveSpan(tracing_1.ClickHouseSpanNames.query, {
            kind: tracing_1.ClickHouseSpanKind.CLIENT,
            attributes: this.withBaseSpanAttributes({
                "clickhouse.response.format": format,
                "clickhouse.request.query_id": queryParams.query_id,
                "clickhouse.request.session_id": queryParams.session_id,
            }),
        }, async (span) => {
            let queryResult;
            try {
                queryResult = await this.connection.query({
                    query,
                    ...queryParams,
                });
            }
            catch (err) {
                (0, tracing_1.recordSpanError)(span, err);
                span.end();
                throw err;
            }
            const { stream, query_id, response_headers, http_status_code } = queryResult;
            // The query_id may have been generated by the connection layer.
            span.setAttributes({ "clickhouse.request.query_id": query_id });
            setResponseSpanAttributes(span, { http_status_code });
            // The clickhouse.query span covers the HTTP request lifetime only:
            // it ends here, once response headers are received.  A separate
            // clickhouse.query.stream child span is started and handed to the
            // ResultSet to track the response stream consumption (decoded bytes,
            // returned rows, streaming errors).  Separating the two spans makes
            // it easy to distinguish the original request duration from a stream
            // that may never end (e.g. tailing a live table).
            span.end();
            return this.tracer.startActiveSpan(tracing_1.ClickHouseSpanNames.query_stream, { kind: tracing_1.ClickHouseSpanKind.CLIENT }, (streamSpan) => {
                try {
                    return this.makeResultSet(stream, format, query_id, (err) => {
                        if (log_level <= logger_1.ClickHouseLogLevel.ERROR) {
                            log_writer.error({
                                err,
                                module: "Client",
                                message: "Error while processing the ResultSet.",
                                args: {
                                    session_id: queryParams.session_id,
                                    role: queryParams.role,
                                    query_id,
                                },
                            });
                        }
                    }, response_headers, this.jsonHandling, streamSpan);
                }
                catch (err) {
                    (0, tracing_1.recordSpanError)(streamSpan, err);
                    streamSpan.end();
                    throw err;
                }
                // streamSpan is NOT ended here - the ResultSet owns its lifetime.
            });
        });
    }
    /**
     * It should be used for statements that do not have any output,
     * when the format clause is not applicable, or when you are not interested in the response at all.
     * The response stream is destroyed immediately as we do not expect useful information there.
     * Examples of such statements are DDLs or custom inserts.
     *
     * @note if you have a custom query that does not work with {@link ClickHouseClient.query},
     * and you are interested in the response data, consider using {@link ClickHouseClient.exec}.
     */
    async command(params) {
        const query = removeTrailingSemi(params.query.trim());
        const ignore_error_response = params.ignore_error_response ?? false;
        const queryParams = this.withClientQueryParams(params);
        return this.tracer.startActiveSpan(tracing_1.ClickHouseSpanNames.command, {
            kind: tracing_1.ClickHouseSpanKind.CLIENT,
            attributes: this.withBaseSpanAttributes({
                "clickhouse.request.query_id": queryParams.query_id,
                "clickhouse.request.session_id": queryParams.session_id,
            }),
        }, async (span) => {
            try {
                const result = await this.connection.command({
                    query,
                    ignore_error_response,
                    ...queryParams,
                });
                span.setAttributes({
                    "clickhouse.request.query_id": result.query_id,
                });
                setResponseSpanAttributes(span, result);
                return result;
            }
            catch (err) {
                (0, tracing_1.recordSpanError)(span, err);
                throw err;
            }
            finally {
                span.end();
            }
        });
    }
    /**
     * Similar to {@link ClickHouseClient.command}, but for the cases where the output _is expected_,
     * but format clause is not applicable. The caller of this method _must_ consume the stream,
     * as the underlying socket will not be released until then, and the request will eventually be timed out.
     *
     * @note it is not intended to use this method to execute the DDLs, such as `CREATE TABLE` or similar;
     * use {@link ClickHouseClient.command} instead.
     */
    async exec(params) {
        const query = removeTrailingSemi(params.query.trim());
        const values = "values" in params ? params.values : undefined;
        const decompress_response_stream = params.decompress_response_stream ?? true;
        const ignore_error_response = params.ignore_error_response ?? false;
        const queryParams = this.withClientQueryParams(params);
        return this.tracer.startActiveSpan(tracing_1.ClickHouseSpanNames.exec, {
            kind: tracing_1.ClickHouseSpanKind.CLIENT,
            attributes: this.withBaseSpanAttributes({
                "clickhouse.request.query_id": queryParams.query_id,
                "clickhouse.request.session_id": queryParams.session_id,
            }),
        }, async (span) => {
            try {
                const result = await this.connection.exec({
                    query,
                    values,
                    decompress_response_stream,
                    ignore_error_response,
                    ...queryParams,
                });
                span.setAttributes({
                    "clickhouse.request.query_id": result.query_id,
                });
                setResponseSpanAttributes(span, result);
                return result;
            }
            catch (err) {
                (0, tracing_1.recordSpanError)(span, err);
                throw err;
            }
            finally {
                span.end();
            }
        });
    }
    /**
     * The primary method for data insertion. It is recommended to avoid arrays in case of large inserts
     * to reduce application memory consumption and consider streaming for most of such use cases.
     * As the insert operation does not provide any output, the response stream is immediately destroyed.
     *
     * @note in case of a custom insert operation (e.g., `INSERT FROM SELECT`),
     * consider using {@link ClickHouseClient.command}, passing the entire raw query there
     * (including the `FORMAT` clause).
     */
    async insert(params) {
        if (Array.isArray(params.values) && params.values.length === 0) {
            return { executed: false, query_id: "", response_headers: {} };
        }
        const format = params.format || "JSONCompactEachRow";
        this.valuesEncoder.validateInsertValues(params.values, format);
        const query = getInsertQuery(params, format);
        const queryParams = this.withClientQueryParams(params);
        return this.tracer.startActiveSpan(tracing_1.ClickHouseSpanNames.insert, {
            kind: tracing_1.ClickHouseSpanKind.CLIENT,
            attributes: this.withBaseSpanAttributes({
                "db.operation.name": "INSERT",
                "db.collection.name": params.table,
                "clickhouse.request.format": format,
                "clickhouse.request.query_id": queryParams.query_id,
                "clickhouse.request.session_id": queryParams.session_id,
                // Only known up front for array-based inserts; for streamed
                // inserts, the row count is not observable by the client.
                "clickhouse.request.sent_rows": Array.isArray(params.values)
                    ? params.values.length
                    : undefined,
            }),
        }, async (span) => {
            try {
                const values = this.valuesEncoder.encodeValues(params.values, format);
                // TODO: record `clickhouse.request.encoded_bytes` (the
                // pre-compression request body size) here. This is a
                // post-common-deprecation feature: once `client-common` is
                // deprecated and the Node.js / Web clients are fully independent,
                // each client can measure the encoded payload using its own
                // platform-native byte-length primitive (e.g. `Buffer.byteLength`)
                // instead of a hand-rolled common implementation.
                const result = await this.connection.insert({
                    query,
                    values,
                    ...queryParams,
                });
                span.setAttributes({
                    "clickhouse.request.query_id": result.query_id,
                });
                setResponseSpanAttributes(span, result);
                return { ...result, executed: true };
            }
            catch (err) {
                (0, tracing_1.recordSpanError)(span, err);
                throw err;
            }
            finally {
                span.end();
            }
        });
    }
    /**
     * A health-check request. It does not throw if an error occurs - the error is returned inside the result object.
     *
     * By default, Node.js version uses the built-in `/ping` endpoint, which does not verify credentials.
     * Optionally, it can be switched to a `SELECT` query (see {@link PingParamsWithSelectQuery}).
     * In that case, the server will verify the credentials.
     *
     * **NOTE**: Since the `/ping` endpoint does not support CORS, the Web version always uses a `SELECT` query.
     */
    async ping(params) {
        const select = params?.select ?? false;
        return this.tracer.startActiveSpan(tracing_1.ClickHouseSpanNames.ping, {
            kind: tracing_1.ClickHouseSpanKind.CLIENT,
            attributes: this.withBaseSpanAttributes({
                "clickhouse.ping.select": select,
            }),
        }, async (span) => {
            try {
                return await this.connection.ping(params ?? { select: false });
            }
            catch (err) {
                (0, tracing_1.recordSpanError)(span, err);
                throw err;
            }
            finally {
                span.end();
            }
        });
    }
    /**
     * Shuts down the underlying connection.
     * This method should ideally be called only once per application lifecycle,
     * for example, during the graceful shutdown phase.
     */
    async close() {
        return await this.connection.close();
    }
    /**
     * Closes the client connection.
     *
     * Automatically called when using `using` statement in supported environments.
     * @see {@link ClickHouseClient.close}
     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/using
     */
    async [Symbol.asyncDispose]() {
        await this.close();
    }
    withBaseSpanAttributes(extra) {
        const url = this.connectionParams.url;
        const attrs = {
            "db.system.name": "clickhouse",
            "server.address": url.hostname,
            "server.port": getServerPort(url),
            "db.namespace": this.connectionParams.database,
        };
        if (this.connectionParams.application_id !== undefined) {
            attrs["clickhouse.application"] = this.connectionParams.application_id;
        }
        for (const [k, v] of Object.entries(extra)) {
            if (v !== undefined)
                attrs[k] = v;
        }
        return attrs;
    }
    withClientQueryParams(params) {
        return {
            clickhouse_settings: {
                ...this.clientClickHouseSettings,
                ...params.clickhouse_settings,
            },
            query_params: params.query_params,
            abort_signal: params.abort_signal,
            query_id: params.query_id,
            session_id: params.session_id ?? this.sessionId,
            role: params.role ?? this.role,
            auth: params.auth,
            http_headers: params.http_headers,
            use_multipart_params: params.use_multipart_params,
            use_multipart_params_auto: params.use_multipart_params_auto,
        };
    }
}
exports.ClickHouseClient = ClickHouseClient;
function getServerPort(url) {
    if (url.port !== "") {
        return Number(url.port);
    }
    return url.protocol === "https:" ? 443 : 80;
}
/** Records HTTP status and `X-ClickHouse-Summary` counters on the span once
 *  the response (headers) arrived. The summary values are complete only when
 *  the query was executed with `wait_end_of_query=1`; see
 *  {@link ClickHouseSummary}. */
function setResponseSpanAttributes(span, result) {
    const attributes = {};
    if (result.http_status_code !== undefined) {
        attributes["db.response.status_code"] = result.http_status_code;
    }
    const summary = result.summary;
    if (summary !== undefined) {
        attributes["clickhouse.summary.read_rows"] = summary.read_rows;
        attributes["clickhouse.summary.read_bytes"] = summary.read_bytes;
        attributes["clickhouse.summary.written_rows"] = summary.written_rows;
        attributes["clickhouse.summary.written_bytes"] = summary.written_bytes;
        attributes["clickhouse.summary.result_rows"] = summary.result_rows;
        attributes["clickhouse.summary.result_bytes"] = summary.result_bytes;
        attributes["clickhouse.summary.elapsed_ns"] = summary.elapsed_ns;
    }
    span.setAttributes(attributes);
}
function formatQuery(query, format) {
    query = query.trim();
    query = removeTrailingSemi(query);
    return query + " \nFORMAT " + format;
}
function removeTrailingSemi(query) {
    let lastNonSemiIdx = query.length;
    for (let i = lastNonSemiIdx; i > 0; i--) {
        if (query[i - 1] !== ";") {
            lastNonSemiIdx = i;
            break;
        }
    }
    if (lastNonSemiIdx !== query.length) {
        return query.slice(0, lastNonSemiIdx);
    }
    return query;
}
function isInsertColumnsExcept(obj) {
    return (obj !== undefined &&
        obj !== null &&
        typeof obj === "object" &&
        // Avoiding ESLint no-prototype-builtins error
        Object.prototype.hasOwnProperty.call(obj, "except"));
}
function getInsertQuery(params, format) {
    let columnsPart = "";
    if (params.columns !== undefined) {
        if (Array.isArray(params.columns) && params.columns.length > 0) {
            columnsPart = ` (${params.columns.join(", ")})`;
        }
        else if (isInsertColumnsExcept(params.columns) &&
            params.columns.except.length > 0) {
            columnsPart = ` (* EXCEPT (${params.columns.except.join(", ")}))`;
        }
    }
    return `INSERT INTO ${params.table.trim()}${columnsPart} FORMAT ${format}`;
}
