// @clickhouse/client-common/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClickHouseLogLevel = exports.LogWriter = exports.DefaultLogger = void 0;
class DefaultLogger {
    trace({ module, message, args }) {
        const params = [
            formatMessage({ module, message, level: "TRACE" }),
        ];
        if (args) {
            params.push("\nArguments:", args);
        }
        console.debug(...params);
    }
    debug({ module, message, args }) {
        const params = [
            formatMessage({ module, message, level: "DEBUG" }),
        ];
        if (args) {
            params.push("\nArguments:", args);
        }
        console.debug(...params);
    }
    info({ module, message, args }) {
        const params = [
            formatMessage({ module, message, level: "INFO" }),
        ];
        if (args) {
            params.push("\nArguments:", args);
        }
        console.info(...params);
    }
    warn({ module, message, args, err }) {
        const params = [
            formatMessage({ module, message, level: "WARN" }),
        ];
        if (args) {
            params.push("\nArguments:", args);
        }
        if (err) {
            params.push("\nCaused by:", err);
        }
        console.warn(...params);
    }
    error({ module, message, args, err }) {
        const params = [
            formatMessage({ module, message, level: "ERROR" }),
        ];
        if (args) {
            params.push("\nArguments:", args);
        }
        params.push("\nCaused by:", err);
        console.error(...params);
    }
}
exports.DefaultLogger = DefaultLogger;
class LogWriter {
    logger;
    module;
    logLevel;
    constructor(logger, module, logLevel) {
        this.logger = logger;
        this.module = module;
        this.logLevel = logLevel;
        this.info({
            message: `Log level is set to ${exports.ClickHouseLogLevel[this.logLevel]}`,
        });
    }
    trace(params) {
        if (this.logLevel <= exports.ClickHouseLogLevel.TRACE) {
            this.logger.trace({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    debug(params) {
        if (this.logLevel <= exports.ClickHouseLogLevel.DEBUG) {
            this.logger.debug({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    info(params) {
        if (this.logLevel <= exports.ClickHouseLogLevel.INFO) {
            this.logger.info({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    warn(params) {
        if (this.logLevel <= exports.ClickHouseLogLevel.WARN) {
            this.logger.warn({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    error(params) {
        if (this.logLevel <= exports.ClickHouseLogLevel.ERROR) {
            this.logger.error({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
}
exports.LogWriter = LogWriter;
/**
 * Mimics the runtime shape of a numeric TypeScript `enum`: an object with both
 * forward (`Name -> number`) and reverse (`number -> Name`) mappings. This
 * preserves backwards compatibility with the previous `enum ClickHouseLogLevel`
 * — both `ClickHouseLogLevel.TRACE` (returns `0`) and `ClickHouseLogLevel[0]`
 * (returns `'TRACE'`) continue to work — while staying compatible with the
 * `--erasableSyntaxOnly` TypeScript option (no TS `enum` declaration).
 */
exports.ClickHouseLogLevel = {
    /**
     * A fine-grained debugging event. Might produce a lot of logs, so use with caution.
     */
    TRACE: 0,
    /**
     * A debugging event. Useful for debugging, but generally not needed in production. Includes technical values that might require redacting.
     */
    DEBUG: 1,
    /**
     * An informational event. Indicates that an event happened.
     */
    INFO: 2,
    /**
     * A warning event. Not an error, but is likely more important than an informational event. Addressing should help prevent potential issues.
     */
    WARN: 3,
    /**
     * An error event. Something went wrong.
     */
    ERROR: 4,
    /**
     * Logging is turned off.
     */
    OFF: 127,
    0: "TRACE",
    1: "DEBUG",
    2: "INFO",
    3: "WARN",
    4: "ERROR",
    127: "OFF",
};
function formatMessage({ level, module, message, }) {
    const ts = new Date().toISOString();
    return `[${ts}][${level}][@clickhouse/client][${module}] ${message}`;
}
