// @clickhouse/client-common/dist/tracing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopClickHouseTracer = exports.NoopClickHouseSpan = exports.ClickHouseSpanNames = exports.ClickHouseSpanKind = exports.ClickHouseSpanStatusCode = void 0;
exports.recordSpanError = recordSpanError;
const error_1 = require("./error");
/** Value-identical to OTEL's `SpanStatusCode`, so non-OTEL implementations
 *  do not have to deal with magic numbers. */
exports.ClickHouseSpanStatusCode = {
    UNSET: 0,
    OK: 1,
    ERROR: 2,
};
/** Value-identical to OTEL's `SpanKind`. The client only ever uses
 *  {@link ClickHouseSpanKind.CLIENT}. */
exports.ClickHouseSpanKind = {
    INTERNAL: 0,
    SERVER: 1,
    CLIENT: 2,
    PRODUCER: 3,
    CONSUMER: 4,
};
/** Span name constants used by the client when starting spans.
 *  Exposed so that adapters and tests can match on them. */
exports.ClickHouseSpanNames = {
    query: "clickhouse.query",
    /** A child of {@link ClickHouseSpanNames.query} that covers the lifetime
     *  of the `ResultSet` stream - from the first byte read to full
     *  consumption, cancellation, or failure.  Ends with
     *  `clickhouse.response.decoded_bytes` and (for row-streaming paths)
     *  `db.response.returned_rows`. */
    query_stream: "clickhouse.query.stream",
    command: "clickhouse.command",
    exec: "clickhouse.exec",
    insert: "clickhouse.insert",
    ping: "clickhouse.ping",
};
const noop = () => undefined;
/** Shared no-op span handed out by {@link NoopClickHouseTracer}. @internal */
exports.NoopClickHouseSpan = {
    setAttributes: noop,
    setStatus: noop,
    recordException: noop,
    end: noop,
};
/** No-op tracer assigned once at client creation when no tracer is
 *  configured, so the hot path stays branch-free (monomorphic call sites
 *  that the JIT can inline). @internal */
exports.NoopClickHouseTracer = {
    startActiveSpan: (_name, _options, fn) => fn(exports.NoopClickHouseSpan),
};
/** Records the exception on the span and marks it with the ERROR status,
 *  normalizing non-`Error` throwables to `Error`.
 *
 *  Sets the {@link https://opentelemetry.io/docs/specs/semconv/registry/attributes/error/#error-type `error.type`}
 *  attribute to the error class name (e.g. `ClickHouseError`, `TypeError`),
 *  and, for server-side errors ({@link ClickHouseError}), the numeric server
 *  error code as `clickhouse.error.code`. */
function recordSpanError(span, err) {
    const error = err instanceof Error ? err : new Error(String(err));
    const attributes = {
        "error.type": error.constructor.name,
    };
    if (error instanceof error_1.ClickHouseError) {
        const code = Number(error.code);
        attributes["clickhouse.error.code"] = Number.isNaN(code)
            ? error.code
            : code;
    }
    span.setAttributes(attributes);
    span.recordException(error);
    span.setStatus({
        code: exports.ClickHouseSpanStatusCode.ERROR,
        message: error.message,
    });
}
