// @clickhouse/client/dist/connection/compression.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decompressResponse = decompressResponse;
exports.isDecompressionError = isDecompressionError;
exports.createRequestCompressor = createRequestCompressor;
const client_common_1 = require("@clickhouse/client-common");
const stream_1 = __importDefault(require("stream"));
const zlib_1 = __importDefault(require("zlib"));
function decompressResponse(response, log_writer, log_level) {
    const encoding = response.headers["content-encoding"];
    // No `Content-Encoding`: nothing to decompress.
    if (encoding === undefined) {
        return { response };
    }
    if (encoding === "gzip") {
        return {
            response: stream_1.default.pipeline(response, zlib_1.default.createGunzip(), function pipelineCb(err) {
                if (err && log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                    log_writer.error({
                        message: "An error occurred while decompressing the response",
                        err,
                    });
                }
            }),
        };
    }
    if (encoding === "zstd") {
        // zstd's zlib API was added in Node.js 22.15.0; on an older runtime the
        // codec is recognized but unusable, so report that explicitly rather than
        // falling through to a misleading "Unexpected encoding".
        if (typeof zlib_1.default.createZstdDecompress !== "function") {
            return {
                error: new Error("Received a zstd-compressed response, but this Node.js runtime (v" +
                    process.versions.node +
                    ") does not support zstd decompression: the built-in zlib module " +
                    "does not provide the zstd APIs (added in Node.js 22.15.0). Use gzip " +
                    "compression instead, or upgrade Node.js."),
            };
        }
        return {
            response: stream_1.default.pipeline(response, zlib_1.default.createZstdDecompress(), function pipelineCb(err) {
                if (err && log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                    log_writer.error({
                        message: "An error occurred while decompressing the response",
                        err,
                    });
                }
            }),
        };
    }
    if (encoding === "br") {
        return {
            response: stream_1.default.pipeline(response, zlib_1.default.createBrotliDecompress(), function pipelineCb(err) {
                if (err && log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                    log_writer.error({
                        message: "An error occurred while decompressing the response",
                        err,
                    });
                }
            }),
        };
    }
    return { error: new Error(`Unexpected encoding: ${encoding}`) };
}
function isDecompressionError(result) {
    return result.error !== undefined;
}
/** Returns the request-body compressor for the given config. Each codec carries
 *  its own optional tuning option - a `level` for gzip/zstd, a `quality` for
 *  Brotli; when omitted, the codec default is used (Brotli defaults to quality 4
 *  rather than zlib's slow default of 11, which is impractical for a streaming
 *  insert body). The exhaustive switch fails the build if a codec is added to
 *  {@link RequestCompression} without a corresponding case here. */
function createRequestCompressor(compression) {
    switch (compression.codec) {
        case "gzip":
            return zlib_1.default.createGzip(compression.level !== undefined
                ? { level: compression.level }
                : undefined);
        case "zstd":
            return zlib_1.default.createZstdCompress(compression.level !== undefined
                ? {
                    params: {
                        [zlib_1.default.constants.ZSTD_c_compressionLevel]: compression.level,
                    },
                }
                : undefined);
        case "br":
            return zlib_1.default.createBrotliCompress({
                params: {
                    [zlib_1.default.constants.BROTLI_PARAM_QUALITY]: compression.quality ?? 4,
                },
            });
        default: {
            const exhaustiveCheck = compression;
            throw new Error(`Unsupported request compression codec: ${String(exhaustiveCheck.codec)}`);
        }
    }
}
