// @clickhouse/client/dist/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeConfigImpl = void 0;
const client_common_1 = require("@clickhouse/client-common");
const zlib_1 = __importDefault(require("zlib"));
const connection_1 = require("./connection");
const result_set_1 = require("./result_set");
const utils_1 = require("./utils");
function unknownCodecError(value, direction) {
    return new Error(`Unknown ${direction} compression codec "${value}". ` +
        `Supported codecs: gzip, zstd, br.`);
}
// zstd's zlib APIs were added in Node.js 22.15.0, and compression /
// decompression are separate functions - so each direction is gated on its own,
// naming the missing function to make the diagnosis concrete.
/** Fails fast at client creation on an unknown request codec, or on `zstd` when
 *  this Node.js runtime's `zlib` does not provide the zstd compression API -
 *  rather than a TypeError deep in a later insert. The boolean forms (`false`
 *  off, `true` gzip) need no validation and are filtered out by the caller. */
function ensureRequestCodecSupported(value) {
    if (value === "zstd") {
        if (typeof zlib_1.default.createZstdCompress !== "function") {
            throw new Error("zstd compression is not supported by this Node.js runtime (v" +
                process.versions.node +
                "): the built-in zlib module does not provide `createZstdCompress` " +
                "(the zstd APIs were added in Node.js 22.15.0). Use gzip compression instead.");
        }
    }
    else if (value !== "gzip" && value !== "br") {
        throw unknownCodecError(value, "request");
    }
}
/** Fails fast at client creation on an unknown response codec, or on `zstd`
 *  when this Node.js runtime's `zlib` does not provide the zstd decompression
 *  API - rather than a TypeError deep in a later query. The boolean forms
 *  (`false` off, `true` gzip) need no validation and are filtered by the caller. */
function ensureResponseCodecSupported(value) {
    if (value === "zstd") {
        if (typeof zlib_1.default.createZstdDecompress !== "function") {
            throw new Error("zstd compression is not supported by this Node.js runtime (v" +
                process.versions.node +
                "): the built-in zlib module does not provide `createZstdDecompress` " +
                "(the zstd APIs were added in Node.js 22.15.0). Use gzip compression instead.");
        }
    }
    else if (value !== "gzip" && value !== "br") {
        throw unknownCodecError(value, "response");
    }
}
exports.NodeConfigImpl = {
    handle_specific_url_params: (config, url) => {
        const nodeConfig = { ...config };
        const unknownParams = new Set();
        const handledParams = new Set();
        const urlSearchParamsKeys = [...url.searchParams.keys()];
        if (urlSearchParamsKeys.length > 0) {
            urlSearchParamsKeys.forEach((key) => {
                const value = url.searchParams.get(key);
                switch (key) {
                    case "keep_alive_idle_socket_ttl":
                        if (nodeConfig.keep_alive === undefined) {
                            nodeConfig.keep_alive = {};
                        }
                        nodeConfig.keep_alive.idle_socket_ttl = (0, client_common_1.numberConfigURLValue)({
                            key,
                            value,
                            min: 0,
                        });
                        handledParams.add(key);
                        break;
                    default:
                        unknownParams.add(key);
                }
            });
        }
        return {
            config: nodeConfig,
            unknown_params: unknownParams,
            handled_params: handledParams,
        };
    },
    make_connection: (nodeConfig, params) => {
        const { compress_request, decompress_response } = params.compression;
        if (compress_request) {
            ensureRequestCodecSupported(compress_request.codec);
        }
        if (decompress_response) {
            ensureResponseCodecSupported(decompress_response.codec);
        }
        let tls = undefined;
        if (nodeConfig.tls !== undefined) {
            if ("cert" in nodeConfig.tls && "key" in nodeConfig.tls) {
                tls = {
                    type: "Mutual",
                    ...nodeConfig.tls,
                };
            }
            else {
                tls = {
                    type: "Basic",
                    ...nodeConfig.tls,
                };
            }
        }
        // normally, it should be already set after processing the config
        const keep_alive = {
            enabled: nodeConfig?.keep_alive?.enabled ?? true,
            idle_socket_ttl: nodeConfig?.keep_alive?.idle_socket_ttl ?? 2500,
        };
        return connection_1.NodeConnectionFactory.create({
            connection_params: params,
            set_basic_auth_header: nodeConfig.set_basic_auth_header ?? true,
            capture_enhanced_stack_trace: nodeConfig.capture_enhanced_stack_trace ?? false,
            eagerly_destroy_stale_sockets: nodeConfig.keep_alive?.eagerly_destroy_stale_sockets ?? false,
            http_agent: nodeConfig.http_agent,
            keep_alive,
            tls,
            max_response_headers_size: nodeConfig.max_response_headers_size,
        });
    },
    values_encoder: (jsonHandling) => new utils_1.NodeValuesEncoder(jsonHandling),
    make_result_set: ((stream, format, query_id, log_error, response_headers, jsonHandling, span) => result_set_1.ResultSet.instance({
        stream,
        format,
        query_id,
        log_error,
        response_headers,
        jsonHandling,
        span,
    })),
};
