// @clickhouse/client/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultJSONHandling = exports.ClickHouseSpanKind = exports.ClickHouseSpanStatusCode = exports.ClickHouseSpanNames = exports.TupleParam = exports.isException = exports.isRow = exports.isProgressRow = exports.SimpleColumnTypes = exports.parseColumnType = exports.RecordsJSONFormats = exports.SingleDocumentJSONFormats = exports.StreamableJSONFormats = exports.StreamableFormats = exports.SupportedRawFormats = exports.SupportedJSONFormats = exports.SettingsMap = exports.ClickHouseLogLevel = exports.parseError = exports.ClickHouseError = exports.drainStream = exports.ResultSet = exports.createClient = exports.ClickHouseClient = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "ClickHouseClient", { enumerable: true, get: function () { return client_1.NodeClickHouseClient; } });
var client_2 = require("./client");
Object.defineProperty(exports, "createClient", { enumerable: true, get: function () { return client_2.createClient; } });
var result_set_1 = require("./result_set");
Object.defineProperty(exports, "ResultSet", { enumerable: true, get: function () { return result_set_1.ResultSet; } });
var stream_1 = require("./connection/stream");
Object.defineProperty(exports, "drainStream", { enumerable: true, get: function () { return stream_1.drainStream; } });
/**
 * Re-export @clickhouse/client-common runtime values.
 *
 * These are intentionally re-exported through local bindings (rather than a direct
 * `export { ... } from '@clickhouse/client-common'`) so that the `@deprecated` JSDoc tags
 * applied to them in `@clickhouse/client-common` are NOT propagated to consumers of this package.
 * Importing these values from `@clickhouse/client` is the recommended, non-deprecated path.
 */
const client_common_1 = require("@clickhouse/client-common");
exports.ClickHouseError = client_common_1.ClickHouseError;
exports.parseError = client_common_1.parseError;
exports.ClickHouseLogLevel = client_common_1.ClickHouseLogLevel;
exports.SettingsMap = client_common_1.SettingsMap;
exports.SupportedJSONFormats = client_common_1.SupportedJSONFormats;
exports.SupportedRawFormats = client_common_1.SupportedRawFormats;
exports.StreamableFormats = client_common_1.StreamableFormats;
exports.StreamableJSONFormats = client_common_1.StreamableJSONFormats;
exports.SingleDocumentJSONFormats = client_common_1.SingleDocumentJSONFormats;
exports.RecordsJSONFormats = client_common_1.RecordsJSONFormats;
exports.parseColumnType = client_common_1.parseColumnType;
exports.SimpleColumnTypes = client_common_1.SimpleColumnTypes;
exports.isProgressRow = client_common_1.isProgressRow;
exports.isRow = client_common_1.isRow;
exports.isException = client_common_1.isException;
exports.TupleParam = client_common_1.TupleParam;
exports.ClickHouseSpanNames = client_common_1.ClickHouseSpanNames;
exports.ClickHouseSpanStatusCode = client_common_1.ClickHouseSpanStatusCode;
exports.ClickHouseSpanKind = client_common_1.ClickHouseSpanKind;
exports.defaultJSONHandling = client_common_1.defaultJSONHandling;
