// @clickhouse/client/dist/result_set.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResultSet = void 0;
const client_common_1 = require("@clickhouse/client-common");
const client_common_2 = require("@clickhouse/client-common");
const buffer_1 = require("buffer");
const stream_1 = __importStar(require("stream"));
const utils_1 = require("./utils");
const NEWLINE = 0x0a;
class ResultSet {
    response_headers = {};
    exceptionTag = undefined;
    log_error;
    jsonHandling;
    _consumed = false;
    /**
     * The stream of the response body.
     *
     * It is expected that the stream is passed directly from the response of the HTTP request
     * and has not been consumed or altered yet.
     */
    _stream;
    format;
    /** The `clickhouse.query.stream` span owned by this result set (if the
     *  client was configured with a tracer); it ends via {@link finishSpan}
     *  when the response stream is fully consumed, closed, or fails. */
    span;
    /** Decoded (decompressed) bytes received from the server so far. */
    span_bytes = 0;
    /** Rows decoded from the response stream so far. */
    span_rows = 0;
    span_rows_counted = false;
    span_finished = false;
    query_id;
    constructor(_stream, format, query_id, log_error, _response_headers, jsonHandling, span) {
        this._stream = _stream;
        this.format = format;
        this.query_id = query_id;
        this.span = span;
        this.jsonHandling = {
            ...client_common_1.defaultJSONHandling,
            ...jsonHandling,
        };
        // eslint-disable-next-line no-console
        this.log_error = log_error ?? ((err) => console.error(err));
        if (_response_headers !== undefined) {
            this.response_headers = Object.freeze(_response_headers);
            this.exceptionTag = _response_headers[client_common_1.EXCEPTION_TAG_HEADER_NAME];
        }
    }
    consume() {
        if (this._consumed) {
            throw new Error(streamAlreadyConsumedMessage);
        }
        this._consumed = true;
        return this._stream;
    }
    /** Add the number of rows decoded from the response stream. */
    addSpanRows(count) {
        this.span_rows_counted = true;
        this.span_rows += count;
    }
    /** Record the final response metrics (`clickhouse.response.decoded_bytes`
     *  and, when rows were counted, `db.response.returned_rows`) and the error
     *  (if any) on the span, and end it. Safe to call multiple times - only
     *  the first call wins. */
    finishSpan(err) {
        if (!this.span || this.span_finished) {
            return;
        }
        this.span_finished = true;
        const attributes = {
            "clickhouse.response.decoded_bytes": this.span_bytes,
        };
        if (this.span_rows_counted) {
            attributes["db.response.returned_rows"] = this.span_rows;
        }
        this.span.setAttributes(attributes);
        if (err) {
            (0, client_common_1.recordSpanError)(this.span, err);
        }
        this.span.end();
    }
    /** See {@link BaseResultSet.text}. */
    async text() {
        const stream = this.consume();
        try {
            const text = await (0, utils_1.getAsText)(stream);
            this.span_bytes += buffer_1.Buffer.byteLength(text);
            this.finishSpan();
            return text;
        }
        catch (err) {
            this.finishSpan(err);
            throw err;
        }
    }
    /** See {@link BaseResultSet.json}. */
    async json() {
        // JSONEachRow, etc.
        if ((0, client_common_2.isStreamableJSONFamily)(this.format)) {
            const result = [];
            // Using the stream() instead of _stream directly to leverage the existing logic
            // for handling incomplete chunks and exception tags.
            // The span progress is updated and the span is finished by the stream() pipeline.
            // TODO: consider using stream() for all formats to unify the logic and error handling.
            const stream = this.stream();
            for await (const rows of stream) {
                for (const row of rows) {
                    result.push(row.json());
                }
            }
            return result;
        }
        // JSON, JSONObjectEachRow, etc.
        if ((0, client_common_2.isNotStreamableJSONFamily)(this.format)) {
            const stream = this.consume();
            try {
                const text = await (0, utils_1.getAsText)(stream);
                this.span_bytes += buffer_1.Buffer.byteLength(text);
                this.finishSpan();
                return this.jsonHandling.parse(text);
            }
            catch (err) {
                this.finishSpan(err);
                throw err;
            }
        }
        // should not be called for CSV, etc.
        throw new Error(`Cannot decode ${this.format} as JSON`);
    }
    /** See {@link BaseResultSet.stream}. */
    stream() {
        (0, client_common_2.validateStreamFormat)(this.format);
        const incompleteChunks = [];
        const logError = this.log_error;
        const exceptionTag = this.exceptionTag;
        const jsonHandling = this.jsonHandling;
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const resultSet = this;
        const toRows = new stream_1.Transform({
            transform(chunk, _encoding, callback) {
                resultSet.span_bytes += chunk.length;
                const rows = [];
                let lastIdx = 0;
                let currentChunkPart;
                while (true) {
                    // an unescaped newline character denotes the end of a row,
                    // or at least the beginning of the exception marker
                    const idx = chunk.indexOf(NEWLINE, lastIdx);
                    if (idx === -1) {
                        incompleteChunks.push(chunk.subarray(lastIdx));
                        if (rows.length > 0) {
                            resultSet.addSpanRows(rows.length);
                            this.push(rows);
                        }
                        break;
                    }
                    else {
                        // Check for exception in the chunk (only after 25.11)
                        if (exceptionTag !== undefined &&
                            idx >= 1 &&
                            chunk[idx - 1] === client_common_1.CARET_RETURN) {
                            return callback((0, client_common_1.extractErrorAtTheEndOfChunk)(chunk, exceptionTag));
                        }
                        if (incompleteChunks.length > 0) {
                            incompleteChunks.push(chunk.subarray(lastIdx, idx));
                            currentChunkPart = buffer_1.Buffer.concat(incompleteChunks);
                            // Removing used buffers and reusing the already allocated memory
                            // by setting length to 0
                            incompleteChunks.length = 0;
                        }
                        else {
                            currentChunkPart = chunk.subarray(lastIdx, idx);
                        }
                        const text = currentChunkPart.toString();
                        rows.push({
                            text,
                            json() {
                                return jsonHandling.parse(text);
                            },
                        });
                        lastIdx = idx + 1; // skipping newline character
                    }
                }
                callback();
            },
            autoDestroy: true,
            objectMode: true,
        });
        const pipeline = stream_1.default.pipeline(this.consume(), toRows, function pipelineCb(err) {
            if (err &&
                err.name !== "AbortError" &&
                err.message !== resultSetClosedMessage) {
                logError(err);
                resultSet.finishSpan(err);
            }
            else {
                resultSet.finishSpan();
            }
        });
        return pipeline;
    }
    /** See {@link BaseResultSet.close}. */
    close() {
        // NB: _consumed is intentionally NOT set here. The stream is destroyed
        // (not read to completion), so marking it as "consumed" would produce the
        // misleading "Stream has been already consumed" error on any subsequent
        // call. Downstream code will instead receive a "ResultSet has been closed"
        // error from the destroyed stream, which accurately describes what happened.
        this._stream.destroy(new Error(resultSetClosedMessage));
        this.finishSpan();
    }
    /**
     * Closes the `ResultSet`.
     *
     * Automatically called when using `using` statement in supported environments.
     * @see {@link ResultSet.close}
     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/using
     */
    [Symbol.dispose]() {
        this.close();
    }
    static instance({ stream, format, query_id, log_error, response_headers, jsonHandling, span, }) {
        return new ResultSet(stream, format, query_id, log_error, response_headers, jsonHandling, span);
    }
}
exports.ResultSet = ResultSet;
const streamAlreadyConsumedMessage = "Stream has been already consumed";
const resultSetClosedMessage = "ResultSet has been closed";
